entitlement/src/main/java/org/killbill/billing/entitlement/api/DefaultEntitlementApi.java

```diff
@@ -350,6 +340,37 @@ public Entitlement doCall(final EntitlementApi entitlementApi, final Entitlement
         return pluginExecution.executeWithPlugin(addEntitlementWithPlugin, pluginContext);
     }
 
+    private void preCheckAddEntitlement(final UUID bundleId, final DateTime entitlementRequestedDate, final BaseEntitlementWithAddOnsSpecifier baseEntitlementWithAddOnsSpecifier, final CallContext callContext) throws EntitlementApiException {
+        final InternalCallContext context = internalCallContextFactory.createInternalCallContext(bundleId, ObjectType.BUNDLE, callContext);
+        final List<SubscriptionBase> subscriptionsByBundle;
+
+        try {
+            subscriptionsByBundle = subscriptionBaseInternalApi.getSubscriptionsForBundle(bundleId, null, context);
+
+            if (subscriptionsByBundle == null || subscriptionsByBundle.size() == 0) {
+                throw new EntitlementApiException(ErrorCode.SUB_NO_ACTIVE_SUBSCRIPTIONS, bundleId);
+            }
+        } catch (SubscriptionBaseApiException e) {
+            throw new EntitlementApiException(e);
+        }
+
+        if (!subscriptionsByBundle.get(0).getCategory().equals(ProductCategory.STANDALONE)) {
```

Comment:
Nit: flip `equals()`.

Reply:
I would suggest to harden this check:
1. If the subscription we are trying to add is a `STANDALONE`, existing bundle should only contain `STANDALONE` subscriptions -- maybe this is enough to check the first one
2.  If the subscription we are trying to add is a `ADD_ON`, existing bundle should one `BASE` -- and potentially other `ADD_ON`, we could skip this validation part. In this case, the existing logic around cancellation should remain.

Reply:
@wwjbatista I am not really following, i think this is fine. I'll take a look once the code is merged, and potentially tweak things if needed.